In [52]:
import pandas as pd

df = pd.read_csv('../data/e-commerce.csv', encoding='latin-1')


In [53]:
print(df.shape)


(541909, 8)


#Answer 541909 rows and 8 columns



In [54]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  str    
 1   StockCode    541909 non-null  str    
 2   Description  540455 non-null  str    
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  str    
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  str    
dtypes: float64(2), int64(1), str(5)
memory usage: 33.1 MB


For datatype there the columns Customer ID and InvoiceDate should in str and datetime respectively.
For I have spot out some missing infos let me spot out.

In [55]:
df.isna().sum() # Missing records 
#1454 records in description are missing
#135080 of customer ID are missing


InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135080
Country             0
dtype: int64

In [56]:
df.describe()

,Quantity,UnitPrice,CustomerID
count,541909.000000,541909.000000,406829.000000
mean,9.552250,4.611114,15287.690570
std,218.081158,96.759853,1713.600303
min,-80995.000000,-11062.060000,12346.000000
25%,1.000000,1.250000,13953.000000
50%,3.000000,2.080000,15152.000000
75%,10.000000,4.130000,16791.000000
max,80995.000000,38970.000000,18287.000000


Quantity has outlier in both upper and lower which needs data analytics and there should not be any negative quantity numbers
Unitprice also has outlier in both bounds and negative is an anaomaly

In [57]:
"""df['InvoiceDate'] = pd.to_datetime(df['InvoiceDate'], format='%m-%d-%Y %H:%M')
print(df['InvoiceDate'].dtype)
print(df['InvoiceDate'].head())"""

"df['InvoiceDate'] = pd.to_datetime(df['InvoiceDate'], format='%m-%d-%Y %H:%M')\nprint(df['InvoiceDate'].dtype)\nprint(df['InvoiceDate'].head())"

In [58]:
df['InvoiceDate'].head()

0    12/1/2010 8:26
1    12/1/2010 8:26
2    12/1/2010 8:26
3    12/1/2010 8:26
4    12/1/2010 8:26
Name: InvoiceDate, dtype: str

In [59]:
df['CustomerID'].astype('str').head()

0    17850.0
1    17850.0
2    17850.0
3    17850.0
4    17850.0
Name: CustomerID, dtype: str

In [60]:
print(df.head())
#p

  InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom  


print the first 5 records index 0 to 4
what is invoice no 
what is inovice date
unitprice means the e.g the white metal lantern sold at 3.39 pounds
both invoice no. is the same as customer id?
the customer id 17850 has ordered these things

In [61]:
df[df['CustomerID'] == 17850]['InvoiceNo'].nunique()

35

In [62]:
df['CustomerID'].nunique()

4372

In [63]:
No_of_transactions_in_each_country=df.groupby('Country')['InvoiceNo'].nunique()
No_of_transactions_in_each_country

Country
Australia                  69
Austria                    19
Bahrain                     4
Belgium                   119
Brazil                      1
Canada                      6
Channel Islands            33
Cyprus                     20
Czech Republic              5
Denmark                    21
EIRE                      360
European Community          5
Finland                    48
France                    461
Germany                   603
Greece                      6
Hong Kong                  15
Iceland                     7
Israel                      9
Italy                      55
Japan                      28
Lebanon                     1
Lithuania                   4
Malta                      10
Netherlands               101
Norway                     40
Poland                     24
Portugal                   71
RSA                         1
Saudi Arabia                2
Singapore                  10
Spain                     105
Sweden                     46
Sw

In [64]:
No_of_transactions_in_each_country.idxmax()


'United Kingdom'

In [65]:
df['Revenue']=df['Quantity']*df['UnitPrice']
#Revenue gain from each record

In [66]:
#total revenue

sum=df['Revenue'].sum()
sum

np.float64(9747747.934)

In [71]:
df['Sales'] = df['Quantity'] * df['UnitPrice']
neg = df[(df['Quantity'] < 0) | (df['UnitPrice'] < 0)]
neg['Sales'].sum()

np.float64(-918936.61)

In [68]:
df.groupby('Description')['Quantity'].sum().idxmax()

'WORLD WAR 2 GLIDERS ASSTD DESIGNS'